services/process-dependencies

```diff
@@ -0,0 +1,227 @@
+  let depWrap = false;
+  manifest.split('\n').forEach((line) => {
+    if ((depWrap) && (!line.startsWith(' '))) {
+      depWrap = false;
+    }
+
+    const matches = line.match(/^Plugin-Dependencies: (.*)?/);
+
+    if (matches) {
+      dependencies.push(matches[1])
+      depWrap = true;
+    }
+    else if (depWrap) {
+      dependencies.push(line.trim());
+    }
+  });
+
+  dependencies = dependencies.join('').split(',');
+
+  // Returning an array of dependency tuples [artifactId, version]
+  return dependencies
+    .filter(entry => !entry.match(/\=optional/))
+    .map(entry => entry.split(':'));
+};
+
+/*
+ * Fetch a specified plugin manifest
+ *
+ * @return Promise
+ */
+const fetchPluginManifest = async (pluginUrl) => {
+  try {
+    let start = Date.now();
+    const response = await request(`${pluginUrl}!META-INF/MANIFEST.MF`);
+    logger.info(`Fetching ${pluginUrl} took ${Date.now() - start}`);
+    return processManifest(response);
+  } catch (err) {
+    if (err.statusCode == 404) {
+      logger.error(`Could not find ${pluginUrl}`);
+    }
+    else {
+      logger.warn(err);
+    }
+    return null;
+  }
+};
+
+/*
+ * processPlugin will generate the appropriate ingest metadata for the given
+ * plugin based on its version
+ */
+const processPlugin = (plugin, into) => {
+  logger.info('Processing', plugin.artifactId);
+
+  const pluginFilename = path.join(plugin.artifactId, plugin.version, `${plugin.artifactId}-${plugin.version}.hpi`);
+  const groupPath = plugin.groupId.replace(/\./g, '/');
+  let url = `${RELEASES}${groupPath}/`;
+
+  if (plugin.version.match(/(.*?)-rc(\d+)\.(.*)?/)) {
+    logger.debug(`${plugin.artifactId} looks incremental`);
+    url = `${INCREMENTALS}${groupPath}/`;
+  }
+  url = url + pluginFilename;
+
+  let record = {
+    groupId: plugin.groupId,
+    artifactId: plugin.artifactId,
+    url: url,
+    checksum: {},
+  };
+
+  into.push(record);
+
+  tasks.push(
+    request({
+        method: 'HEAD',
+        uri: url,
+        resolveWithFullResponse: true,
+    }).then((response) => {
+      Object.assign(record.checksum, {
+          type: 'sha256',
+          signature: response.headers['x-checksum-sha256'],
+      });
+    })
+  );
+
+  return fetchPluginManifest(url).then(d => d);
+};
+
+// The plugin name map will help us later determine what has been specified in
+// the ssentials.yaml compared to an implied dependency
+const pluginNameMap = new Map(essentials.spec.plugins.map(entry => [entry.artifactId, entry]));
+// Perform a first pass on the plugins needed for all Evergreen installations
+const firstPass = essentials.spec.plugins.map(plugin => processPlugin(plugin, ingest.plugins));
+
+// Collection of plugins which are needed but not specified in essentials.yaml
+let needed = {};
+
+
+const resolvePlugin = (name, inventory) => {
+  // We don't have the plugin in essentials.yaml, so it's dependencies
+  // should be resolved through the update center unless otherwise
```

Comment:
This is a bad policy—nondeterministic! Use the least upper bound instead.

Reply:
Please help me understand why pulling the latest released version of plugins is not a good thing for the mission of Jenkins Essentials here.

@batmat the desire for testing you describe is still easily done with providing a mock update center during test.

We can chat more about the resolution behavior in the meeting this morning

Reply:
> not dependent on new releases in UC

In particular I see no reason why the script should load the UC JSON at all. It should be irrelevant.

Ideally it would use the Artifactory API to parse manifests except in cases where the plugin already exists in the local filesystem in which case it can more efficiently load that.

Reply:
Some options described by @jglick 

1. Make `essentials.yaml` contain every single plugin in the graph
1. Load in transitive dependencies by collecting the listing of depX and select the highest version specified.

Another option: use route 1 with some automation to check for updates, e.g. `workflow-job`, and submit a pull request to `essentials.yaml`